common.py

```diff
@@ -403,21 +403,29 @@ def prepare_filelist(self, names, path, goto, indent):
     def is_hidden(self, filename, path, goto=''):
         if not (path or goto):  # special case for ThisPC
             return False
-        tests = self.view.settings().get('dired_hidden_files_patterns', ['.*'])
-        if isinstance(tests, str):
-            tests = [tests]
-        if any(fnmatch.fnmatch(filename, pattern) for pattern in tests):
-            return True
-        if sublime.platform() != 'windows':
-            return False
-        # check for attribute on windows:
-        try:
-            attrs = ctypes.windll.kernel32.GetFileAttributesW(join(path, goto, filename))
-            assert attrs != -1
-            result = bool(attrs & 2)
-        except (AttributeError, AssertionError):
-            result = False
-        return result
+        show_hidden = self.show_hidden
+        show_excluded = self.view.settings().get('dired_show_excluded_files', True)
+        is_hidden = False
+        if not show_hidden:
+            tests = self.view.settings().get('dired_hidden_files_patterns', ['.*'])
+            if isinstance(tests, str):
+                tests = [tests]
+            if any(fnmatch.fnmatch(filename, p) for p in tests):
+                is_hidden = True
+            if sublime.platform() == 'windows':
+                # check for attribute on windows:
+                try:
+                    attrs = ctypes.windll.kernel32.GetFileAttributesW(join(path, goto, filename))
+                    assert attrs != -1
+                    if bool(attrs & 2):
+                        is_hidden = True
+                except:
+                    pass
+        if not show_excluded:
+            tests = self.view.settings().get('folder_exclude_patterns', []) + self.view.settings().get('file_exclude_patterns', [])
```

Comment:
Yea, you see the problem is that mixes folders and files which is wrong because folder pattern may match files which are not supposed to be excluded and vice versa.
So then we have to check if name isdir twice, which seems lame.
I dunno, maybe add new method which would check if hidden and add os.sep for folders then call this method instead of `items = [name for name in os.listdir(path) if not self.is_hidden(name, path)]` in appropriate places;
that way we would check for isdir once and call `is_hidden` with additional argument `isdir=True/False`.
What your thoughts are?

Also, the check for attribute on Windows must be the last one, because it is the heaviest so we should not execute it if name was already defined as hidden by other check.